File: avalon/merlin/api/src/java/org/apache/avalon/merlin/Kernel.java
/* 
 * Copyright 2004 Apache Software Foundation
 * Licensed  under the  Apache License,  Version 2.0  (the "License");
 * you may not use  this file  except in  compliance with the License.
 * You may obtain a copy of the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed  under the  License is distributed on an "AS IS" BASIS,
 * WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
 * implied.
 * 
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avalon.merlin;

import org.apache.avalon.merlin.event.KernelEventListener;
import org.apache.avalon.activation.appliance.Appliance;
import org.apache.avalon.activation.appliance.Block;

/**
 * A Kernel is the root of a containment solution. This interfaces 
 * defines the contract for any kernel implementation covering 
 * management aspects and service resolution aspects.
 *
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.4 $ $Date: 2004/01/24 23:25:31 $
 */
public interface Kernel extends KernelController
{
    static final int INITIALIZING = 0;
    static final int INITIALIZED = 1;
    static final int STARTING = 2;
    static final int ASSEMBLY = 3;
    static final int DEPLOYMENT = 4;
    static final int STARTED = 5;
    static final int STOPPING = 6;
    static final int DECOMMISSIONING = 7;
    static final int DISSASSEMBLY = 8;
    static final int STOPPED = 9;

   /**
    * Add a kernel listener.
    * @param listener the kernel listener to be added
    */
    void addKernelEventListener( KernelEventListener listener );

   /**
    * Remove a kernel listener.
    * @param listener the kernel listener to be removed
    */
    void removeKernelEventListener( KernelEventListener listener );

   /**
    * Return the root block.
    * @return the root application containment block
    */
    Block getBlock();

   /**
    * Return the applicance matching the supplied path.
    * @return the appliance
    */
    Appliance locate( String path ) throws KernelException;

   /**
    * Return the current state of the kernel.
    * @return the kernel state
    */
    int getState();

}

File: avalon-sandbox/avalon-net/Castle/MicroKernel/Kernel.cs
// Copyright 2004 The Apache Software Foundation
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

namespace Apache.Avalon.Castle.MicroKernel
{
	using System;

	public delegate void DependencyListenerDelegate( Type service, IHandler handler );

	/// <summary>
	/// Defines the Kernel service
	/// </summary>
	public interface Kernel
	{
		/// <summary>
		/// Adds a component to kernel.
		/// </summary>
		/// <param name="key">The unique key that identifies the component</param>
		/// <param name="service">The service exposed by this component</param>
		/// <param name="implementation">The actual implementation</param>
		void AddComponent( String key, Type service, Type implementation );

		/// <summary>
		/// Adds an aspect to kernel.
		/// </summary>
		/// <param name="flags">May be Before, After or both.</param>
		/// <param name="aspect">The aspect implementation</param>
		void AddAspect( AspectPointCutFlags flags, IAspect aspect );

		/// <summary>
		/// Obtains a component handler using the 
		/// unique component key
		/// </summary>
		IHandler this [ String key ]
		{
			get;
		}

		/// <summary>
		/// Gets or Sets the IHandlerFactory implementation
		/// </summary>
		IHandlerFactory HandlerFactory
		{
			get;
			set;
		}

		/// <summary>
		/// Gets or Sets the ILifestyleManagerFactory implementation
		/// </summary>
		ILifestyleManagerFactory LifestyleManagerFactory
		{
			get;
			set;
		}

		/// <summary>
		/// Returns an array of aspects 
		/// interested in a specific point cut
		/// </summary>
		/// <param name="pointcut"></param>
		/// <returns></returns>
		IAspect[] GetAspects( AspectPointCutFlags pointcut );

		/// <summary>
		/// Returns true if kernel "knows" the specified 
		/// service
		/// </summary>
		/// <param name="service">The service interface</param>
		/// <returns>true if is already registered</returns>
		bool HasService( Type service );

		/// <summary>
		/// Used by handlers to register itself as 
		/// and dependency to be satisfied.
		/// </summary>
		/// <param name="service">The service interface</param>
		/// <param name="depDelegate">Delegate to be invoked</param>
		void AddDependencyListener( Type service, DependencyListenerDelegate depDelegate );

		/// <summary>
		/// Returns a IHandler implementation for 
		/// the specified service
		/// </summary>
		/// <param name="service">The service interface</param>
		/// <returns>IHandler implementation</returns>
		IHandler GetHandlerForService( Type service );
	}
}
